src/Store/Features.re

```diff
@@ -186,6 +186,25 @@ let update =
     | None => (state, Effect.none)
     }
 
+  | FileDrop(msg) =>
+    let eff =
+      switch (msg) {
+      | FilesDroppedOnEditor({paths}) =>
+        Isolinear.Effect.createWithDispatch(
+          ~name="feature.filedrop.openFiles", dispatch => {
+          List.iter(
+            path => {
+              let stats = Unix.stat(path);
+              if (stats.st_kind == S_REG) {
+                dispatch(OpenFileByPath(path, None, None));
+              };
+            },
+            paths,
+          )
+        })
```

Comment:
Ir would be nice to have this effect moved out of the update function, to avoid it turning into a blob. I also think ti does too much. Ideally this would be two separate effects, one for the `stat`, which ought to be non-blocking and dispatch a new user-supplied message when done, and then another to open the file. That would also enable the stat effect to be moved into the feature project, while the open file effect would need to be an `outmsg` because it's hard-coded to `Actions.t` for historic reasons.

This area is still a bit murky with a lot of historical baggage and few good examples of how to do these things properly, so don't worry too much about it, but the general lesson here is that supplying the message to be dispatched when it is invoked allows effects to be generalized.

Reply:
Here's one example of such an effect:

https://github.com/onivim/oni2/blob/d971dae8cb82807a310744bfca2b995ecdfad929/src/Store/FileExplorerStore.re#L11-L25

Used here:

https://github.com/onivim/oni2/blob/d971dae8cb82807a310744bfca2b995ecdfad929/src/Store/FileExplorerStore.re#L52-L59

And another:

https://github.com/onivim/oni2/blob/d971dae8cb82807a310744bfca2b995ecdfad929/src/Feature/SCM/Feature_SCM.re#L65-L74

Used here:

https://github.com/onivim/oni2/blob/d971dae8cb82807a310744bfca2b995ecdfad929/src/Store/ExtensionClientStoreConnector.re#L246-L249

Reply:
@bryphe and I worked this out last night and I think the consensus we came to was that eventually the drag and drop feature will have it's own state once the _drag_ part is implemented. I.e. if you held a dragged file over a certain area for x amount of time, something happens, etc. I'm obviously not as familiar with the underlying architecture as the two of you are, so it's possible @bryphe is able to explain better than I am.

I understand what you mean with regards to separating the actions -- I can do that. Thanks for the examples!